torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1531,7 +1592,13 @@ void ProcessGroupNCCL::watchdogHandler() {
     for (auto it = workMetaList_.begin(); it != workMetaList_.end();
          /* no increment */) {
       auto& work = *it;
-      work.checkAndSetException();
+      // When terminateProcessGroup_ is true, communicators have already been
+      // aborted, So cannot check exception based on them. But watchdog needs to
```

Comment:
This seems to apply to any method of WorkNCCL because any method can touch ncclComm_. Would be nice to figure out a holistic solution. For example, if `terminateProcessGroup_.load()` is true, can this thread just return?

Reply:
Return was my first implementation, however, it has conflicts with some existing assumption that the thread would continue to run to finish already enqueued work checks, e.g., dec_pending_event_queries